tests/v2/validation/upgrade/kubernetes_test.go

```diff
@@ -50,142 +39,57 @@ func (u *UpgradeKubernetesTestSuite) SetupSuite() {
 	clusters, err := upgradeinput.LoadUpgradeKubernetesConfig(client)
 	require.NoError(u.T(), err)
 
-	require.NotEmptyf(u.T(), clusters, "couldn't generate the config for the upgrade test")
 	u.clusters = clusters
 }
 
 func (u *UpgradeKubernetesTestSuite) TestUpgradeKubernetes() {
-	for _, cluster := range u.clusters {
-		cluster := cluster
-		if cluster.Name == local {
-			u.Run(cluster.Name, func() {
-				if cluster.IsUpgradeDisabled {
-					u.T().Skipf("Kubernetes upgrade is disabled for [%v]", cluster.Name)
-				}
+	tests := []struct {
+		name   string
+		client *rancher.Client
+	}{
+		{"Upgrade ", u.client},
+	}
 
-				u.testUpgradeLocalCluster(cluster.Name, cluster.VersionToUpgrade, cluster.IsLatestVersion)
-			})
-		} else {
-			u.Run(cluster.Name, func() {
-				if cluster.IsUpgradeDisabled {
-					u.T().Skipf("Kubernetes upgrade is disabled for [%v]", cluster.Name)
+	for _, tt := range tests {
+		for _, cluster := range u.clusters {
+			testName := tt.name
+			if cluster.Name == local {
+				testName += "Local cluster to " + cluster.VersionToUpgrade
+				u.Run(tt.name, func() {
+					upgradeLocalCluster(u.T(), tt.client, cluster.Name, cluster.VersionToUpgrade, cluster.IsLatestVersion)
+				})
+			} else {
+				clusterObject, _, _ := clusters.GetProvisioningClusterByName(tt.client, cluster.Name, namespace)
+				if clusterObject == nil {
+					testName += "RKE1 cluster to " + cluster.VersionToUpgrade
+				} else {
```

Comment:
I know its probably hard to avoid but, there is too much nested logic here all back-to-back. It reads as complicated in the test, which is why for fresh install provisioning we have permutations to abstract this away from the test a bit. Might not be worth pulling this into a similar function yet, but just noting it could be improved.

Reply:
Let's speak offline on this. There are some architectural things that we might want to address that can potentially resolve this.

Reply:
Resolved